boards: frdm_mcux947_mcxn947_cpu0_qspi: enable mcuboot by default

hakehuang opened the pull request:
frdm_mcux947_mcxn947_cpu0_qspi needs MCUboot to be programmed to internal flash, which will load and execute the Zephyr application from QSPI.

butok commented:
I am not against this particular PR.

This qspi is confusing and problematic: https://github.com/zephyrproject-rtos/zephyr/pull/72905#issuecomment-2358255535
Let's think if we can make it better.
For example https://github.com/nxphovergames/doom-mcx  is using external flash without Mcuboot. + @PetervdPerk-NXP

hakehuang commented:
> For example https://github.com/nxphovergames/doom-mcx is using external flash without Mcuboot. + @PetervdPerk-NXP

https://github.com/NXPHoverGames/Doom-MCX/blob/d4fad3e4afed85297f5853e7a4d49ed024fbd6e3/boards/frdm_mcxn947.ld#L8C5-L8C81 is it from the internal flash?

butok commented:
> > For example https://github.com/nxphovergames/doom-mcx is using external flash without Mcuboot. + @PetervdPerk-NXP
> 
> https://github.com/NXPHoverGames/Doom-MCX/blob/d4fad3e4afed85297f5853e7a4d49ed024fbd6e3/boards/frdm_mcxn947.ld#L8C5-L8C81 is it from the internal flash?

```
[268/268] Linking C executable zephyr\zephyr.elf
c:/users/nxa12205/zephyr-sdk-0.16.8/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd.exe:zephyr/linker.cmd:277: warning: memory region `IDT_LIST' not declared
Memory region         Used Size  Region Size  %age Used
           FLASH:      372464 B         2 MB     17.76%
             RAM:      393912 B       416 KB     92.47%
  EXTERNAL_FLASH:     3842112 B         4 MB     91.60%
           SRAM1:          0 GB        96 KB      0.00%
       *default*:   268807920 B 18446744073709551615 B      0.00%
Generating files from C:/_ddm/ZEPHYR/doom-mcx-workspace/doom-mcx/build/zephyr/zephyr.elf for board: frdm_mcxn947
```

PetervdPerk-NXP commented:
@hakehuang 
https://github.com/NXPHoverGames/Doom-MCX/
Is using both internal and external flash.
Program and Zephyr kernel is on internal flash.
Assets are on external flash through a custom linker script.
https://github.com/NXPHoverGames/Doom-MCX/blob/d4fad3e4afed85297f5853e7a4d49ed024fbd6e3/boards/frdm_mcxn947_mcxn947_cpu0.conf#L33
https://github.com/NXPHoverGames/Doom-MCX/blob/d4fad3e4afed85297f5853e7a4d49ed024fbd6e3/boards/frdm_mcxn947.ld#L23

Although I don't think Doom-MCX has any problems with this PR, it has it's own defconfig and linkerscript.

hakehuang committed:
boards: frdm_mcux947_mcxn947_cpu0_qspi: enable mcuboot by default

frdm_mcux947_mcxn947_cpu0_qspi needs MCUboot to be programmed
to internal flash, which will load and execute from QSPI.

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

Commit:
samples: canopennode.no_storage: add filter for sample

add not CONFIG_STREAM_FLASH filter

Signed-off-by: Hake Huang <hake.huang@oss.nxo.com>

nordicjm commented on boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0_qspi_defconfig:
```diff
@@ -16,3 +16,5 @@ CONFIG_HW_STACK_PROTECTION=y
 
 # Enable TrustZone-M
 CONFIG_TRUSTED_EXECUTION_SECURE=y
+CONFIG_MCUBOOT_SIGNATURE_KEY_FILE="bootloader/mcuboot/root-rsa-2048.pem"
```
this is the default so it can go

nordicjm reviewed (commented):
any particular reason you haven't added a sysbuild file to enable mcuboot?

hakehuang commented:
> any particular reason you haven't added a sysbuild file to enable mcuboot?

for this platform, the mcuboot needs burn to internal flash for once. as the mcuboot is in the internal-flash and  now the coding is running in external nor flash, so we do not need to do this twice.

nordicjm commented:
> > any particular reason you haven't added a sysbuild file to enable mcuboot?
> 
> for this platform, the mcuboot needs burn to internal flash for once. as the mcuboot is in the internal-flash and now the coding is running in external nor flash, so we do not need to do this twice. besides, Is there an easy to add sysbuild for mcuboot option at board level? instead to adding for each application? Thanks in advance. @nordicjm

Yes, see https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/dptechnics/walter/Kconfig.sysbuild (the signature type bit can be omitted)

butok commented:
Hi all

FYI
**The mcxn947 Boot mode can be controlled with the CMPA[BOOT_SOURCE] bit of the PFR**:
- CMPA[BOOT_SOURCE]= 01b the boot ROM will run FlexSPI NOR boot handling routine.
- CMPA[BOOT_SOURCE]= 00b or 11b internal flash memory handling routine is executed (default state).

It means that MCUboot can be putted to an external flash, or an application may run in external flash without MCUboot.

So the frdm_mcux947_mcxn947_cpu0_qspi should be redone and not mandate using of MCUboot in the internal flash as it is done now. 
What do you think?  + @mmahadevan108

henrikbrixandersen commented on samples/modules/canopennode/sample.yaml:
```diff
@@ -29,4 +29,7 @@ tests:
       SB_CONFIG_BOOTLOADER_MCUBOOT=y
 
   sample.modules.canopennode.no_storage:
+    filter: not CONFIG_BOOTLOADER_MCUBOOT
     extra_args: CONF_FILE=prj_no_storage.conf
+    platform_exclude:
+      - frdm_mcxn947/mcxn947/cpu0/qspi
```
This change does not match the Git commit message. Please explain why this is needed.

hakehuang replied:
if use no_storage conf, the frdm_mcxn947/mcxn947/cpu0/qspi can not support this test instance. as it is runs from qspi.

butok commented:
The frdm_mcux947_mcxn947_cpu0_qspi should be removed.
It's confusing and buggy.

henrikbrixandersen commented:
> @henrikbrixandersen @hakehuang have you two resolved the CR?

Yes, but the review comment from @nordicjm still isn't addressed (and the PR is still in draft).

hakehuang commented:
> Yes, but the review comment from @nordicjm still isn't addressed (and the PR is still in draft).

I already enable as @nordicjm  suggested. Thanks

hakehuang commented:
> https://github.com/NXPHoverGames/Doom-MCX/blob/d4fad3e4afed85297f5853e7a4d49ed024fbd6e3/boards/frdm_mcxn947.ld#L23

I see only use external flash as data, right?

DerekSnell commented:
Just wrapping up this discussion, I finally submitted https://github.com/zephyrproject-rtos/zephyr/pull/90409 to boot directly from QSPI without requiring a secondary bootloader.
